api/document/serializers/doc_cursor.py

```diff
@@ -118,15 +118,55 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.context['parent_serializer'] = self
 
+    @property
+    def is_root(self) -> bool:
+        return self.context.get('is_root', True)
+
     def get_meta(self, instance):
         """Include meta data which applies to the whole node."""
         meta = Meta(
             instance,
-            self.context.get('is_root', True),
+            self.is_root,
             self.context.get('policy'),
         )
         return MetaSerializer(meta, context={'parent_serializer': self}).data
 
+    def validate_footnote_citations(self, data: PrimitiveDict,
+                                    footnote_emblems: Set[str]):
+        citations = [
+            util.get_content_text(f['inlines']).strip()
+            for f in util.iter_inlines(data['content'])
+            if f['content_type'] == 'footnote_citation'
+        ]
+
+        for citation in citations:
+            if citation not in footnote_emblems:
+                raise ValidationError(
+                    f"Citation for '{citation}' has no matching footnote"
+                )
+
+    def validate_footnote_emblems(self, data: PrimitiveDict) -> Set[str]:
+        footnote_emblems = [
+            f['type_emblem'] for f in util.iter_children(data['children'])
+            if f['node_type'] == 'footnote'
+        ]
+
+        for emblem in util.iter_non_unique(footnote_emblems):
+            raise ValidationError(
+                f"Multiple footnotes exist with type emblem '{emblem}'"
+            )
+
+        return set(footnote_emblems)
+
+    def to_internal_value(self, data: PrimitiveDict) -> PrimitiveDict:
+        data = super().to_internal_value(data)
+        if data['node_type'] == 'footnote' and not data.get('type_emblem'):
+            raise ValidationError('Footnotes must have type emblems')
+        if self.is_root:
+            footnote_emblems = self.validate_footnote_emblems(data)
+            self.validate_footnote_citations(data, footnote_emblems)
+        return data
```

Comment:
Looks like the [`validate()`](http://www.django-rest-framework.org/api-guide/serializers/#object-level-validation) serializer-level method might be preferable to overriding `to_internal_value`.

Reply:
Oh good point.  So many methods, it gets confusing which one to put stuff in sometimes...

Reply:
I'd argue that sort of unknown inheritance problem is one of the biggest deficits of OO.

Reply:
Yeah! Down with OO!

I like typeclasses and traits, tho...